File: Programmers/KaKao/2019_겨울인턴십_크레인 인형뽑기.java
import java.util.Stack;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        Stack<Integer> stack = new Stack<>();
        
        int[] topIdx = new int[board.length];
        for(int x = 0; x < board.length; x++) {
            for(int y = 0; y < board.length; y++) {
                if(board[y][x] != 0) {
                    topIdx[x] = y;
                    break;
                }
            }
        }
        
        for(int move : moves) {
            if(topIdx[move - 1] == board.length) { continue; }
            
            int tempItem = board[topIdx[move - 1]][move - 1];
            
            if(tempItem != 0) {
                if(!stack.isEmpty() && tempItem == stack.peek()) {
                    stack.pop();
                    board[topIdx[move - 1]][move - 1] = 0;
                    answer += 2;
                }
                else {
                    stack.add(tempItem);
                    board[topIdx[move - 1]][move - 1] = 0;
                }
                
                if(topIdx[move - 1] < board.length) {
                    topIdx[move - 1] = topIdx[move - 1] + 1;
                }
            }
        }
        
        return answer;
    }
}

File: Programmers/KaKao/2019_겨울인턴십_크레인 인형뽑기.kt
import java.util.Stack

class Solution {
    fun solution(board: Array<IntArray>, moves: IntArray): Int {
        var answer = 0

        var bucket: Stack<Int> = Stack()
        val boardSize = board.size

        for(move in moves) {
            var temp = 0
            for(idx in 0 until boardSize) {
                if(board[idx][move - 1] != 0) {
                    temp = board[idx][move - 1]
                    board[idx][move - 1] = 0
                    break
                }
            }

            if(temp == 0) { continue }

            if(!bucket.isEmpty() && bucket.peek() == temp) {
                bucket.pop()
                answer += 2
            }
            else {
                bucket.add(temp)
            }
        }

        return answer
    }
}
